prologin/workernode/operations.py

```diff
@@ -46,308 +50,319 @@ def untar(content, path, compression='gz'):
+        }
+        self.limits = {
+            'wall-time': self.config['timeout'].get('compile', 400),
+            'fsize': 50 * 1024,
+        }
+        self.allowed_dirs = [
+            '/tmp:rw',
+            '/etc',
+            self.config['path']['player_env'],
+        ]
+
+    def write_champion(self, champion_tgz_b64):
+        ctgz = b64decode(champion_tgz_b64)
+        (self.isolator.path / 'champion.tgz').write_bytes(ctgz)
+        return Path('/box')
+
+    def write_compile_script(self):
+        compile_script_path = Path(__file__).parent / 'compile-champion.sh'
+        compile_script = compile_script_path.read_text()
+        compile_script_path = self.isolator.path / 'compile-champion.sh'
+        compile_script_path.write_text(compile_script)
+        compile_script_path.chmod(0o755)
+        return Path('/box/compile-champion.sh')
+
+    def read_champion_compiled(self):
+        compiled_path = self.isolator.path / 'champion-compiled.tar.gz'
+        self.result['champion_compiled'] = b64encode(
+            compiled_path.read_bytes()
+        ).decode()
+
+    async def _run(self, *, champion_tgz_b64):
+        compile_script = self.write_compile_script()
+        champion_path = self.write_champion(champion_tgz_b64)
+
+        cmd = [
+            str(compile_script),
+            self.config['path']['player_env'],
+            str(champion_path),
+        ]
+        await self.isolator.run(cmd, merge_outputs=True)
+
+        self.read_champion_compiled()
+
+
+class SpawnServer(Operation):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.result = {
+            'dump': None,
+            'replay': None,
+            'stats': None,
+            'match_result': None,
+            **self.result,
+        }
+        self.limits = {'wall-time': self.config['timeout'].get('server', 400)}
+        self.allowed_dirs = [
+            '/var',
+            '/etc',
+            '/tmp',
+            os.path.dirname(self.config['path']['stechec_server']),
+            os.path.dirname(self.config['path']['rules']),
+        ]
+
+    def read_compress_b64(self, path):
+        return b64encode(gzip.compress(path.read_bytes())).decode()
+
+    async def _run(
+        self, *, sockets_dir, rep_addr, pub_addr, nb_players, map_content
+    ):
+        self.allowed_dirs.append(sockets_dir + ':rw')
+        # fmt: off
+        cmd = [
+            self.config['path']['stechec_server'],
+            "--rules", self.config['path']['rules'],
+            "--rep_addr", rep_addr,
+            "--pub_addr", pub_addr,
+            "--nb_clients", str(nb_players),
+            "--time", "3000",
+            "--socket_timeout", "45000",
+            "--dump", "/box/dump.json",
+            "--replay", "/box/replay",
+            "--stats", "/box/stats.yaml",
+            "--verbose", "1",
+        ]
+        # fmt: on
+
+        if map_path := self.write_map(map_content):
```

Comment:
W A L R U S

(now you don't have any excuse not to use type annotations when they make sense)